### PR TITLE
chore: Avoid some component constructor code smells

### PR DESCRIPTION
**Pull request by Cervator**
Tested by making sure cities still spawn with working farms and citizens

### CODE REVIEW

**Review by pollend — DISMISSED** [review]

File: src/main/java/org/terasology/metalrenegades/ai/actions/AddSocialComponentAction.java [hunk 1]
```diff
@@ -35,10 +22,10 @@
 
     @Override
     public BehaviorState modify(Actor actor, BehaviorState result) {
-        SimpleSourceComponent socialSourceComponent = new SimpleSourceComponent(CitizenNeed.Type.SOCIAL);
+        SimpleSourceComponent socialSourceComponent = new SimpleSourceComponent();
+        socialSourceComponent.needType = CitizenNeed.Type.SOCIAL;
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/ai/actions/AddSocialComponentAction.java:26]: is it really a bad smell to have additional constructors at all? I was about to introduce new ones for some components :rolling_eyes:  if you're using components programmatically and you need to create them with specific values, each user would end up writing a method to do that in one line ...

For instance, compare https://github.com/Terasology/Xmas/blob/e823584a898480cbd7889b76deafe728144696ff/src/main/java/org/terasology/xmas/player/SurpriseStartingInventory.java#L80-L93

File: src/main/java/org/terasology/metalrenegades/ai/system/FactionAlignmentSystem.java [hunk 2]
```diff
@@ -69,7 +56,8 @@ public Prefab getPrefab(Alignment alignment) {
 
     @ReceiveEvent
     public void onSettlementRegisterEvent(SettlementRegisterEvent buildingEntitySpawnedEvent, EntityRef entityRef) {
-        entityRef.addComponent(new FactionAlignmentComponent(Alignment.values()[(int) (Math.random() * Alignment.values().length)]));
+        FactionAlignmentComponent factionAlignmentComponent = new FactionAlignmentComponent();
+        factionAlignmentComponent.alignment = Alignment.values()[(int) (Math.random() * Alignment.values().length)];
+        entityRef.addComponent(factionAlignmentComponent);
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/ai/system/FactionAlignmentSystem.java:61]: We should only use `addComponent` if we're sure that the component cannot be present already. 

If we want to override the component, you can use `addOrSaveComponent(factionAlignmentComponent)`.

File: src/main/java/org/terasology/metalrenegades/interaction/component/CityCropComponent.java [hunk 3]
```diff
@@ -18,10 +18,4 @@
      * The prefab name of the block form of this crop.
      */
     public String plantName;
-
-    public CityCropComponent(String item, String plant) {
```
Comment (skaldarnar) [inline src/main/java/org/terasology/metalrenegades/interaction/component/CityCropComponent.java]: What about adding an empty constructor instead and still keeping this one?

**Review by skaldarnar — COMMENTED** [review]
I'm really curious how we want to go about ways to construct components programmatically. 

It would be nice to be able to create components with a single line of code (if you have all the field values present), be it by having a constructor to set all fields, some static method, or some (generated) builder-like pattern...

File: src/main/java/org/terasology/metalrenegades/ai/actions/AddSocialComponentAction.java (hunk 1 above)
Comment (pollend) [reply src/main/java/org/terasology/metalrenegades/ai/actions/AddSocialComponentAction.java:26]: constructors define an initial state which is better then just letting people submit anything.  I rather have a constructor then let the state be anything.

**Comment by skaldarnar** [comment]
I only see the constructor as a shorthand notation to create a component with a specific state. I would even go so far to say that:

- there MUST be no constructor at all

or

- there MUST be a constructor with empty arguments, and
- there MAY be a constructor taking an argument for each field with a direct correlation

This restriction would help to avoid putting logic into the components. Either you fully define the state of the component when you create it, or you create the default component state and just do mutations on that.

**Comment by Cervator** [comment]
We discussed this briefly on Discord earlier but just to get a note in here: we definitely could do more on architecture here. Mostly I'm just addressing current state with brief bits of effort expecting a major upheaval when Gestalt v7 lands. Hopefully by then we can codify the conventions using FindBugs or something that can scan for patterns and warn about things we don't want :-)

In the meantime I'm fine either way on either merging this or leaving it alone. I'm not convinced that current state actually causes any bugs, just that similar ones have happened from similar code in the past elsewhere 👍

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**